zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -387,46 +496,294 @@ fn check_orchard_subtrees(db: &ZebraDb) -> bool {
     {
         // Check that there's an entry for every completed orchard subtree root in all orchard trees
         let Some(subtree) = db.orchard_subtree_by_index(index) else {
-            error!(?index, "missing subtree");
-            is_valid = false;
+            result = Err("missing subtree");
+            error!(?result, index);
             continue;
         };
 
         // Check that the subtree end height matches that in the orchard trees.
         if subtree.end != height {
             let is_complete = tree.is_complete_subtree();
-            error!(?subtree.end, ?height, ?index, ?is_complete, "bad orchard subtree end height");
-            is_valid = false;
+            result = Err("bad orchard subtree end height");
+            error!(?result, ?subtree.end, ?height, ?index, ?is_complete, );
         }
 
         // Check the root if the orchard note commitment tree at this height is a complete subtree.
         if let Some((_index, node)) = tree.completed_subtree_index_and_root() {
             if subtree.node != node {
-                error!("completed subtree roots should match");
-                is_valid = false;
+                result = Err("completed subtree roots should match");
+                error!(?result);
             }
         }
     }
 
-    if !is_valid {
+    if result.is_err() {
         error!(
+            ?result,
             ?subtree_count,
-            first_incomplete_subtree_index, "missing or bad orchard subtrees"
+            first_incomplete_subtree_index,
+            "missing or bad orchard subtrees"
+        );
+    }
+
+    result
+}
+
+/// Calculates a note commitment subtree for Sapling, reading blocks from `read_db` if needed.
+///
+/// The subtree must be completed by a note commitment in the block at `end_height`.
+/// `tree` is the tree for that block, and `prev_tree` is the tree for the previous block.
+///
+/// `prev_tree` is only used to rebuild the subtree if it was completed without using the last
+/// note commitment in the block at `end_height`.
+///
+/// # Panics
+///
+/// If a subtree is not completed by a note commitment in the block at `end_height`.
+#[must_use = "subtree should be written to the database after it is calculated"]
+#[instrument(skip(read_db, prev_tree, tree))]
+fn calculate_sapling_subtree(
+    read_db: &ZebraDb,
+    prev_end_height: Height,
+    prev_tree: Arc<sapling::tree::NoteCommitmentTree>,
+    end_height: Height,
+    tree: Arc<sapling::tree::NoteCommitmentTree>,
+) -> NoteCommitmentSubtree<sapling::tree::Node> {
+    // If a subtree is completed by a note commentment in the block at `end_height`,
```

Comment:
###### Optional Suggestion

```suggestion
    // If a subtree is completed by a note commitment in the block at `end_height`,
```